package/src/components/plugins/upload-images.ts

```diff
@@ -0,0 +1,153 @@
+          placeholder.setAttribute("class", "img-placeholder");
+          const image = document.createElement("img");
+          image.setAttribute(
+            "class",
+            "opacity-40 rounded-lg border border-stone-200"
+          );
+          image.src = src;
+          placeholder.appendChild(image);
+          const deco = Decoration.widget(pos + 1, placeholder, {
+            id,
+          });
+          set = set.add(tr.doc, [deco]);
+        } else if (action && action.remove) {
+          set = set.remove(
+            set.find(null, null, (spec) => spec.id == action.remove.id)
+          );
+        }
+        return set;
+      },
+    },
+    props: {
+      decorations(state) {
+        return this.getState(state);
+      },
+    },
+  });
+
+export default UploadImagesPlugin;
+
+function findPlaceholder(state: EditorState, id: {}) {
+  const decos = uploadKey.getState(state);
+  const found = decos.find(null, null, (spec) => spec.id == id);
+  return found.length ? found[0].from : null;
+}
+
+export function startImageUpload(file: File, view: EditorView, pos: number) {
+  // check if the file is an image
+  if (!file.type.includes("image/")) {
+    toast.error("File type not supported.");
+    return;
+
+    // check if the file size is less than 20MB
+  } else if (file.size / 1024 / 1024 > 20) {
+    toast.error("File size too big (max 20MB).");
+    return;
+  }
+
+  // A fresh object to act as the ID for this upload
+  const id = {};
+
+  // Replace the selection with a placeholder
+  const tr = view.state.tr;
+  if (!tr.selection.empty) tr.deleteSelection();
+
+  const reader = new FileReader();
+  reader.readAsDataURL(file);
+  reader.onload = () => {
+    tr.setMeta(uploadKey, {
+      add: {
+        id,
+        pos,
+        src: reader.result,
+      },
+    });
+    view.dispatch(tr);
+  };
+
+  handleImageUpload(file).then((src) => {
+    const { schema } = view.state;
+
+    let pos = findPlaceholder(view.state, id);
+    // If the content around the placeholder has been deleted, drop
+    // the image
+    if (pos == null) return;
+
+    // Otherwise, insert it at the placeholder's position, and remove
+    // the placeholder
+
+    // When BLOB_READ_WRITE_TOKEN is not valid or unavailable, read
+    // the image locally
+    const imageSrc = typeof src === "object" ? reader.result : src;
+
+    const node = schema.nodes.image.create({ src: imageSrc });
+    const transaction = view.state.tr
+      .replaceWith(pos, pos, node)
+      .setMeta(uploadKey, { remove: { id } });
+    view.dispatch(transaction);
+  });
+}
+
+export const handleImageUpload = (file: File) => {
+  // upload to Vercel Blob
+  return new Promise((resolve) => {
+    toast.promise(
+      fetch("/api/upload", {
+        method: "POST",
+        headers: {
+          "content-type": file?.type || "application/octet-stream",
+          "x-vercel-filename": file?.name || "image.png",
```

Comment:
Not able to upload with image name `Screenshot 2023-10-23 at 10.13.48 AM`

https://github.com/naveennaidu/novel-vue/assets/28999143/75764db8-822f-47b2-b435-8b2ad80f236c

You can fix this by using `encodeURIComponent(file?.name || "image.png")`

Reply:
you chan show me **devtools -> network status**

Reply:
Not making any network call. this error because of name encoding issue

Reply:
I got it, and this issue is resolved